### PR TITLE
cicd: build the agent image using github actions

### PR DESCRIPTION
**Pull request by gomex**
Add arm32v7 docker build.

Replace the build from circleCI.

You can check the result on my private hub:

https://hub.docker.com/repository/docker/gomex/agent/tags?page=1&name=765c0c818178dfccf8960b5975ea57ba90fa88bd

Before to merge it, please add these variables:

DOCKER_USERNAME
DOCKER_PASSWORD

PS: Please don't use a personal account.

### CODE REVIEW

File: .github/workflows/build-agent.yml [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        uses: crazy-max/ghaction-docker-buildx@v3
+
+      - name: Set github reference env
+        run: echo ::set-env name=RELEASE_VERSION::${GITHUB_REF#refs/*/}
+      
+      - name: build AMD64 image
+        run: |
+          docker buildx build --push -f agent/Dockerfile.amd64 \
+            --tag shellhubio/agent:${{ github.sha }}-amd64 \
+            --platform linux/amd64 .
```
Comment (otavio) [inline .github/workflows/build-agent.yml:37]: We could unify the architectures here and do it in a for loop, like:

```
for arch in amd64 arm32v6 arm64v8; do
    docker buildx build \
                --push -f agent/Dockerfile.$arch \
                --tag shellhubio/agent:${{ github.sha }}-$arch \
                --platform linux/$arch .
done
```

File: .github/workflows/build-agent.yml [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+      - name: Apply github tag to AMD64 image
+        if: contains(github.ref, 'refs/tags/v')
+        run: |
+          docker pull shellhubio/agent:${{ github.sha }}-amd64 \
+          && docker tag shellhubio/agent:${{ github.sha }}-amd64 shellhubio/agent:${{ env.RELEASE_VERSION }}-amd64 \
+          && docker push shellhubio/agent:${{ env.RELEASE_VERSION }}-amd64
```
Comment (otavio) [inline .github/workflows/build-agent.yml:44]: Here as well.

File: .github/workflows/build-agent.yml [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    - 'agent/**'
+
+  pull_request:
+    branches: [ master ]
+    paths:
+    - 'agent/**'
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
```
Comment (jwillmer) [inline .github/workflows/build-agent.yml:18]: Define a fixed version for your build image or expect that your build will break at some point - been there done that!

Comment (otavio) [reply .github/workflows/build-agent.yml:18]: You mean due to underlying changes on ubuntu image? makes sense.

@gomex can you fix this to 18.04 for now?

Comment (jwillmer) [reply .github/workflows/build-agent.yml:18]: Yes, or sometimes GitHub/GitLab/etc defines different package that are included in the `latest` release

Comment (gomex) [reply .github/workflows/build-agent.yml:18]: I agree ;) I didn't change it yet, but it was in my plan to request it.

File: .github/workflows/build-agent.yml [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+      - name: build multiarch image
+        env:
+          DOCKER_CLI_EXPERIMENTAL: enabled
+        run: |
+            docker manifest create -a shellhubio/agent:${{ github.sha }} shellhubio/agent:${{ github.sha }}-amd64 shellhubio/agent:${{ github.sha }}-arm32v6 shellhubio/agent:${{ github.sha }}-arm32v7 shellhubio/agent:${{ github.sha }}-arm64v8 \
+            && docker manifest annotate shellhubio/agent:${{ github.sha }} shellhubio/agent:${{ github.sha }}-amd64 --os linux --arch amd64 \
+            && docker manifest annotate shellhubio/agent:${{ github.sha }} shellhubio/agent:${{ github.sha }}-arm32v6 --os linux --arch arm --variant v6 \
+            && docker manifest annotate shellhubio/agent:${{ github.sha }} shellhubio/agent:${{ github.sha }}-arm32v7 --os linux --arch arm --variant v7 \
+            && docker manifest annotate shellhubio/agent:${{ github.sha }} shellhubio/agent:${{ github.sha }}-arm64v8 --os linux --arch arm64 --variant v8 \
+            && docker manifest push shellhubio/agent:${{ github.sha }}
```
[the remark this answers is not in the document]
Comment (gomex) [reply .github/workflows/build-agent.yml]: Done

File: .github/workflows/build-agent.yml [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+          arch: arm/6
+        - dockerfile: arm32v7
+          arch: arm/7
+        - dockerfile: arm64v8
+          arch: arm64/8
+
+    steps:
+      - name: checkout code
+        uses: actions/checkout@v2
+
```
Comment (otavio) [inline .github/workflows/build-agent.yml:38]: Please drop one of the new lines.

File: .github/workflows/build-agent.yml [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+            variant=(6 7 8)
+            
+            for ((i = 0; i < 3; i++))  ; do
+              echo "Running ${archname[i]} manifest annotation"
+              docker manifest annotate shellhubio/agent:${{ github.sha }} shellhubio/agent:${{ github.sha }}-${archname[i]} --os linux --arch ${arch[i]} --variant ${variant[i]}
+            done
+            
+            echo "Pushing multiarch manifest"
+            docker manifest push shellhubio/agent:${{ github.sha }}
+
```
Comment (otavio) [inline .github/workflows/build-agent.yml:98]: Two new lines here, please drop one.

Comment (gomex) [reply .github/workflows/build-agent.yml:97]: Done

File: .github/workflows/build-agent.yml (hunk 5 above)
Comment (gomex) [reply .github/workflows/build-agent.yml:38]: Done

**Pull request merged**